### PR TITLE
feat: multi-server librenms_id — track device origin per LibreNMS server

### PR DESCRIPTION
**Pull request by marcinpsk**
## Summary

Replaces the raw integer `librenms_id` custom field with a JSON dict `{"server_key": device_id}` so that devices imported from (or linked to) multiple LibreNMS servers are unambiguously tracked back to their origin. The change is backwards-compatible: legacy bare-integer values continue to work and can be migrated in-place through the UI.

**Depends on PR #244 ** (`pr/librenms-pre-id-multi-server`) — merge that first.
PR #246  (`pr/librems-id-multi-fixes`) stacks on top of this one.

## Motivation / Problem

- Feature

The existing `librenms_id` custom field stores a single integer — the device ID on the LibreNMS server that last touched this NetBox object. This breaks with multiple LibreNMS servers:

- Two servers can independently assign the same integer ID to different physical devices
- There is no way to tell which server an existing link belongs to
- Re-importing from a second server silently overwrites the first server's link

## Scope of Change

- Sync/Import logic
- LibreNMS API interaction
- Config / settings
- Web UI / templates
- Tests

## How Was This Tested?

- Unit tests: yes — 571 tests pass (pytest in devcontainer), covering helpers, views, sync flows, and view wiring
- Manual testing: yes — tested with single and multi-server configs, legacy and new format devices, VC setups

| File | Coverage |
|---|---|
| `test_librenms_id.py` | `get/set/find_by/migrate_legacy` helpers — legacy int passthrough, bool rejection, type-error handling, dict merging, ORM query shape |
| `test_mixins.py` | `CacheMixin.get_cache_key` with and without `server_key` |
| `test_sync_devices.py` | `device_fields.py` sync actions: name, serial, device type, platform, server mapping, legacy ID conversion |
| `test_sync_interfaces.py` | `update_interface_attributes`, `handle_mac_address`, `set_librenms_device_id` path, `None` port_id guard |
| `test_sync_view_mismatch.py` | Mismatch detection logic and `_build_all_server_mappings` ordering / orphaned-server handling |
| `test_view_wiring.py` | Smoke tests — MRO, mixin wiring, permission contracts, template syntax for all views including new ones |

## Risk Assessment

- Does this change affect existing users?
  Yes — but existing devices with bare-integer `librenms_id` continue to work without any action. Migration is explicit and manual via a UI button.
- Could this cause unintended imports / updates?
  No — newly imported devices use the dict format automatically; existing devices are not auto-migrated.

## Backwards Compatibility

- No breaking changes

1. Existing devices with a bare-integer `librenms_id` continue to work — `get_librenms_device_id(obj, "default")` returns the integer for any `server_key`
2. `find_by_librenms_id()` matches both legacy bare-int and new dict formats
3. Migration is **explicit** and **manual** via the Migrate ID format button in the device validation details panel
4. After migration the field shape changes from `42` to `{"default": 42}` (or `{"your_server_key": 42}`)
5. Newly imported devices always use the dict format from the start
6. No Django migration needed — the `librenms_id` custom field type remains `JSONField`; only the stored value shape changes

### New storage format

```
# Legacy (still readable, migrated via explicit UI action)
librenms_id = 42

# New
librenms_id = {"production": 42}
librenms_id = {"production": 42, "lab": 7}   # device visible on both servers
```

All reads/writes go through helpers in `utils.py`:

| Helper | Purpose |
|---|---|
| `get_librenms_device_id(obj, server_key)` | Read ID for a specific server; handles legacy int transparently |
| `set_librenms_device_id(obj, device_id, server_key)` | Write ID; rejects legacy bare-int records (use migration workflow) |
| `find_by_librenms_id(model, id, server_key)` | ORM lookup matching both legacy int and new dict format |
| `migrate_legacy_librenms_id(obj, server_key)` | Explicit one-shot migration; returns True if a migration happened |

### Key changes (delta from PR #244 )

- **`utils.py`** — Four new public ID functions plus `get_librenms_sync_device` with server-aware VC resolution
- **`librenms_api.py`** — `server_key` instance attribute, fallback logic for missing server config
- **`views/mixins.py`** — Optional `server_key` in `CacheMixin.get_cache_key()`
- **`views/sync/device_fields.py`** — `RemoveServerMappingView`, `ConvertLegacyLibreNMSIdView` with collision checks and `select_for_update()` locking
- **`views/base/librenms_sync_view.py`** — `_build_all_server_mappings()` with sorted server entries
- **`views/base/cables_view.py`** — Server-aware cache keys, VC cable verify
- **`views/imports/actions.py`** — Server-aware linking/unlinking, bool rejection in migration pre-checks
- **`templates/librenms_sync_base.html`** — Server card with per-entry Remove button
- **`templates/htmx/device_validation_details.html`** — Per-server badges, Migrate ID format button
- **`urls.py`** — `remove-server-mapping` and `convert-legacy-librenms-id` endpoints

### CODE REVIEW

**Comment by bonzo81** [comment]
Ok @marcinpsk lets looks at getting this one merged. This will be really helpful for multi server instances. As usual I'll kick off Copilot review to see if it finds anything helpful.

File: netbox_librenms_plugin/views/sync/ip_addresses.py [hunk 1]
```diff
@@ -59,14 +65,21 @@ def get_ip_tab_url(self, obj):
             url_name = "plugins:netbox_librenms_plugin:device_librenms_sync"
         else:
             url_name = "plugins:netbox_librenms_plugin:vm_librenms_sync"
-        return f"{reverse(url_name, args=[obj.pk])}?tab=ipaddresses"
+        server_key = getattr(self, "_post_server_key", None) or self.librenms_api.server_key
+        url = f"{reverse(url_name, args=[obj.pk])}?tab=ipaddresses"
+        if server_key:
+            url += f"&server_key={server_key}"
+        return url
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/sync/ip_addresses.py]: `ip_addresses.py`: URL-encode `server_key` in redirect URL

File: netbox_librenms_plugin/views/sync/vlans.py [hunk 2]
```diff
@@ -58,7 +66,11 @@ def _redirect(self, object_type: str, object_id: int):
             if object_type == "device"
             else "plugins:netbox_librenms_plugin:vm_librenms_sync"
         )
-        return redirect(reverse(url_name, kwargs={"pk": object_id}) + "?tab=vlans")
+        server_key = getattr(self, "_post_server_key", None) or self.librenms_api.server_key
+        url = reverse(url_name, kwargs={"pk": object_id}) + "?tab=vlans"
+        if server_key:
+            url += f"&server_key={server_key}"
+        return redirect(url)
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/sync/vlans.py:75]: Fixed in `3f32a46`. Applied `urllib.parse.quote_plus(server_key)` before appending to the redirect URL in `vlans.py`, preventing `&`-injection via a crafted server key.

File: netbox_librenms_plugin/static/netbox_librenms_plugin/js/librenms_sync.js [hunk 3]
```diff
@@ -458,7 +472,12 @@ function verifyVlanInGroup(select, deviceId, vid, vlanType, groupId) {
             vlan_type: vlanType
         })
     })
-        .then(response => response.json())
+        .then(response => {
+            if (!response.ok) {
+                return response.text().then(t => { throw new Error(t || `HTTP ${response.status}`); });
+            }
+            return response.json();
+        })
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/static/netbox_librenms_plugin/js/librenms_sync.js:480]: Fixed in `3f32a46`. Added `.catch(err => { console.error("verifyVlanInGroup error:", err); _vlanVerifyEnd(select); })` to the promise chain so a non-2xx response no longer leaves the modal in a broken/loading state.

File: netbox_librenms_plugin/views/sync/interfaces.py [hunk 4]
```diff
@@ -52,23 +58,27 @@ def post(self, request, object_type, object_id):
...
         interface_name_field = get_interface_name_field(request)
         self.interface_name_field = interface_name_field
         selected_interfaces = self.get_selected_interfaces(request, interface_name_field)
         exclude_columns = request.POST.getlist("exclude_columns")
 
+        redirect_url = (
+            reverse(url_name, kwargs={"pk": object_id})
+            + f"?tab=interfaces&interface_name_field={interface_name_field}"
+            + (f"&server_key={server_key}" if server_key else "")
+        )
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/sync/interfaces.py:76]: Fixed in `3f32a46`. Applied `urllib.parse.quote_plus(server_key)` before appending to the redirect URL in `interfaces.py`.

File: netbox_librenms_plugin/views/sync/cables.py [hunk 5]
```diff
@@ -139,20 +164,23 @@ def post(self, request, pk):
             return error
 
         initial_device = get_object_or_404(Device, pk=pk)
+        server_key = request.POST.get("server_key") or self.librenms_api.server_key
+        self._post_server_key = server_key
+        redirect_url = (
+            f"{reverse('plugins:netbox_librenms_plugin:device_librenms_sync', args=[initial_device.pk])}?tab=cables"
+            + (f"&server_key={server_key}" if server_key else "")
+        )
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/sync/cables.py:203]: Fixed in `3f32a46`. Applied `urllib.parse.quote_plus(server_key)` before appending to the redirect URL in `cables.py`.

File: netbox_librenms_plugin/views/sync/ip_addresses.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_librenms_plugin/views/sync/ip_addresses.py:74]: Fixed in `3f32a46`. Applied `urllib.parse.quote_plus(server_key)` before appending to the redirect URL in `ip_addresses.py`.

**Comment by marcinpsk** [comment]
@bonzo81 it is coming up clean for me other than URL-encode earlier.

**Comment by bonzo81** [comment]
@marcinpsk Finally got some time to look at this
When testing on an existing Netbox dev environment, the 'convert id' button and the device import page I get the same error - ["Invalid value for custom field 'librenms_id': Value must be an integer."]} - . What is your understanding of how this PR handles the existing librenms_id integer values when converting to json? Maybe we need to migrate existing integer librenms_id to json using a Django migration?

**Comment by marcinpsk** [comment]
@bonzo81 yes, it needs migration. I'll check and adjust later - can't access my test setup for next few days.
Nice catch.
should be something like:
```
from django.db import migrations

def _convert_librenms_id_to_json(apps, schema_editor):
    CustomField = apps.get_model("extras", "CustomField")
    try:
        cf = CustomField.objects.get(name="librenms_id")
    except CustomField.DoesNotExist:
        # Custom field hasn't been created yet — nothing to convert.
        return
    if cf.type == "integer":
        cf.type = "json"
        cf.save(update_fields=["type"])

def _revert_librenms_id_to_integer(apps, schema_editor):
    CustomField = apps.get_model("extras", "CustomField")
    try:
        cf = CustomField.objects.get(name="librenms_id")
    except CustomField.DoesNotExist:
        return
    if cf.type == "json":
        cf.type = "integer"
        cf.save(update_fields=["type"])

class Migration(migrations.Migration):
    dependencies = [
        ("extras", "0001_initial"),
        ("netbox_librenms_plugin", "0008_librenmssettings_import_defaults"),
    ]

    operations = [
        migrations.RunPython(
            code=_convert_librenms_id_to_json,
            reverse_code=_revert_librenms_id_to_integer,
        ),
    ]
```

This is handling both legacy and new format - so it doesn't migrate existing values - just a format change.

**Comment by bonzo81** [comment]
> @bonzo81 yes, it needs migration. I'll check and adjust later - can't access my test setup for next few days. Nice catch. should be something like:
> 
> ...
> 

Thanks @marcinpsk

We will need to catch edge cases where the `libenms_id` is a string rather integer. We do this elsewhere in the plugin so essential catch them and migrate them to json too.

**Comment by marcinpsk** [comment]
@bonzo81 Good point on the string edge cases — we're already catching them across all four `librenms_id` helpers in `utils.py`:

- **`get_librenms_device_id()`** — catches bare `str` at top level and inside dict entries; normalises to `int` and auto-saves back
- **`set_librenms_device_id()`** — catches bare `str` at top level; blocks write for digit-strings (same treatment as legacy int), resets non-numeric strings to `{}`
- **`find_by_librenms_id()`** — queries for both `int` and `str(librenms_id)` at the dict-key level and at bare-value level
- **`migrate_legacy_librenms_id()`** — catches digit strings via `isdigit()` and converts to `{server_key: int}`

The `ConvertLegacyLibreNMSIdView` also handles strings explicitly — checks `isdigit()`, casts with `int()`, and re-checks after acquiring the row lock.

I've also pushed a Django migration (`0009_convert_librenms_id_to_json`) that converts the `librenms_id` CustomField type from `integer` to `json` so NetBox validation accepts the new dict format on existing installations.

**Comment by bonzo81** [comment]
Hi @marcinpsk , thanks for update. Away from my dev pc at the moment. Can you confirm that users with the original legacy server config will have compatibility with these changes? The original server config has no server name so do we deny ability to convert, or define a 'default' or 'legacy' server name as part of conversion to new librenms_id?

I think we can merge this just wanted to clarify this point.

**Comment by marcinpsk** [comment]
Hi @bonzo81 
There's convert button, will rewrite ID to JSON format for given device, using active server. No automatic conversion, because it's hard to determine which server it should be.
And yes, all works as right now, if ID is not converted, only field is changed from integer to json, but values are not touched

**Comment by bonzo81** [comment]
Thanks @marcinpsk. Lets merge it!
